File: java/spring/spring-integration/src/main/java/acktsap/spring/resttemplate/converter/ConverterTest.java
package acktsap.spring.resttemplate.converter;

import okhttp3.HttpUrl;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageConverter;
import org.springframework.http.converter.StringHttpMessageConverter;
import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
import org.springframework.web.client.RestTemplate;

import java.util.List;
import java.util.Map;

public class ConverterTest {

    static class TestData {

        private int intValue;

        private String stringValue;

        public void setIntValue(int intValue) {
            this.intValue = intValue;
        }

        public void setStringValue(String stringValue) {
            this.stringValue = stringValue;
        }

        public String toString() {
            return "TestData{" +
                "intValue=" + intValue +
                ", stringValue='" + stringValue + '\'' +
                '}';
        }
    }

    public static void main(String[] args) throws Exception {
        MockWebServer server = new MockWebServer();
        server.start();
        HttpUrl baseUrl = server.url("/test");

        RestTemplate restTemplate = new RestTemplate(getHttpMessageConverters());

        server.enqueue(new MockResponse().addHeader("Content-Type", "text/plain").setBody("""
            {
                "intValue": 3,
                "stringValue": "wow",
                "extraValue": "extra"
            }"""));
        ResponseEntity<TestData> response = restTemplate.getForEntity(baseUrl.uri(), TestData.class);
        System.out.println("body: " + response.getBody());

        restTemplate.close();

        server.shutdown();
    }

    private static List<HttpMessageConverter<?>> getHttpMessageConverters() {
        MappingJackson2HttpMessageConverter mappingJackson2HttpMessageConverter = new MappingJackson2HttpMessageConverter();
        mappingJackson2HttpMessageConverter.setSupportedMediaTypes(List.of(MediaType.TEXT_PLAIN, MediaType.APPLICATION_JSON));
        StringHttpMessageConverter stringHttpMessageConverter = new StringHttpMessageConverter();
        return List.of(
            mappingJackson2HttpMessageConverter,
            stringHttpMessageConverter
        );
    }
}

File: java/spring/spring-integration/src/main/kotlin/acktsap/spring/resttemplate/converter/ConverterTest.kt
package acktsap.spring.resttemplate.converter

import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.springframework.http.MediaType
import org.springframework.http.converter.HttpMessageConverter
import org.springframework.http.converter.StringHttpMessageConverter
import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter
import org.springframework.web.client.RestTemplate
import org.springframework.web.client.postForEntity

fun main(args: Array<String>) {
    val server = MockWebServer().apply {
        start()
    }
    val baseUrl = server.url("/test")

    val restTemplate = RestTemplate(getHttpMessageConverters())
    server.enqueue(MockResponse().addHeader("Content-Type", "application/x-www-form-urlencoded").setBody("""
    {
        "intValue": 3,
        "stringValue": "wow",
        "extraValue": "extra"
    }
    """.trimIndent()))

    val requestBody = mapOf(
        "test" to "value"
    )

    val response = restTemplate.postForEntity<Map<String, String>>(baseUrl.toUri(), requestBody)
    println("body: " + response.body)
    restTemplate.close()

    server.shutdown()
}

private fun getHttpMessageConverters(): List<HttpMessageConverter<*>>? {
    val mappingJackson2HttpMessageConverter = MappingJackson2HttpMessageConverter()
    mappingJackson2HttpMessageConverter.supportedMediaTypes = listOf(
        MediaType.TEXT_PLAIN,
        MediaType.APPLICATION_JSON,
        MediaType.APPLICATION_FORM_URLENCODED,
    )
    val stringHttpMessageConverter = StringHttpMessageConverter()
    return java.util.List.of<HttpMessageConverter<*>>(
        mappingJackson2HttpMessageConverter,
        stringHttpMessageConverter
    )
}
